### DISCUSSION TITLE
"Ghosts" after a deploy

### DISCUSSION THREAD

**Discussion by hbhanoo** [anycable/anycable#193] [Q&A]
I think I have a pretty basic question but couldn't find an answer in the docs.
If I have clients subscribed while I'm doing a deployment, I get this behavior where there ends up being a 'ghost' client.

In my app I want to keep track of each instance of a user joining from multiple tabs, so the connection ID is not tied to a cookie. Instead it's a newly generated UUID for each connection.

```
  class Connection < ActionCable::Connection::Base
    identified_by :current_user, :uuid

    def connect
      self.current_user = find_current_user || guest_user
      self.uuid = "CC-#{SecureRandom.uuid}"
      logger.debug("user #{current_user} joined with #{uuid}")
      reject_unauthorized_connection if current_user.nil?
    end
```

## What I observe: 
My app (like a chat room) shows another instance of the logged in user if the window is open during a deployment

## Hypothesis:

During the deploy, the connection gets dropped without the `unsubscribe` happening on the server side. 

When the user tries to reconnect, they are connected with a new connection ID and the server (which is tracking members of the room in Redis) reports back both the old and the new connection.

## Question:

(Assuming my hypothesis is correct). If I miss the `unsubscribe` call, is there a way to check that my server-side participant list is actually accurate based on current connections? Or is there some other approach I'm missing?

**Comment by palkan** [comment]
> the connection gets dropped without the unsubscribe happening on the server side.

How do you shut down older instances of `anycable-go`? The server tries to make its best to notify RPC servers of disconnects (and, thus, invoke `unsubscribe` callbacks) on shut down (unless killed forcefully).

> is there a way to check that my server-side participant list is actually accurate based on current connections?

In general, it's better to rely on heartbeats than on the callbacks. Callbacks are not reliable; there is always a chance of a server crash.

I shared some presence tracking ideas (and code) here: https://noti.st/palkan/MeBUVe/the-pitfalls-of-realtime-ification (see Resources).

**Reply by hbhanoo** [reply]
> How do you shut down older instances of anycable-go

I'm using heroku for deployment right now. Not entirely sure how it shuts down.

I am currently using a very simple Redis::HashKey to maintain a list of active members for a channel (updated in `subscribe`/`unsubscribe` calls). But it totally makes sense why that's not a smart idea, and I'm digging the code you have at [http://bit.ly/turbo-presence](http://bit.ly/turbo-presence)

**Reply by palkan** [reply]
> I'm using heroku for deployment right now. Not entirely sure how it shuts down.

Heroku let an app to gracefully shutdown for 30s. Can you share the shutdown logs? There should be some lines related to left RPCs calls and the disconnect queue.
